cloudtools/start.py

```diff
@@ -108,40 +106,67 @@ def main(args):
+    init_actions = ['gs://dataproc-initialization-actions/conda/bootstrap-conda.sh',
+                    init_script]
 
     if args.version == 'devel' and args.spark != '2.2.0':
         sys.stderr.write("ERROR: Hail version 'devel' requires Spark 2.2.0.")
         sys.exit(1)
 
     # add VEP init script
     if args.vep:
-        init_actions += ',' + 'gs://hail-common/vep/vep/vep85-init.sh'
+        init_actions.append('gs://hail-common/vep/vep/vep85-init.sh')
 
     # add custom init scripts
     if args.init:
-        init_actions += ',' + args.init
+        init_actions.append(args.init)
 
-    # get Hail build (default to latest)
-    if args.hash == 'latest':
-        hail_hash = check_output(['gsutil', 'cat', 'gs://hail-common/builds/{0}/latest-hash-spark-{1}.txt'.format(args.version, args.spark)]).strip()
+
+    if (not (args.jar and args.zip)) and (args.jar or args.zip):
+        sys.stderr.write('ERROR: pass both --jar and --zip or neither')
+        sys.exit(1)
+
+    jar = None
+    py_zip = None
+
+    if args.jar:
+        jar = args.jar
+        py_zip = args.zip
     else:
-        hail_hash = args.hash
+        if sys.version_info >= (3,0):
+            # Python 3 check_output returns a byte string
+            decode_f = lambda x: x.decode()
+        else:
+            # In Python 2, bytes and str are the same
+            decode_f = lambda x: x
+
+        if args.hash == 'latest':
+            hail_hash = decode_f(check_output(['gsutil', 'cat', 'gs://hail-common/builds/{0}/latest-hash-spark-{1}.txt'.format(args.version, args.spark)]).strip())
+        else:
+            hail_hash = args.hash
+
+        hail_jar = 'hail-{0}-{1}-Spark-{2}.jar'.format(args.version, hail_hash, args.spark)
+        jar = 'gs://hail-common/builds/{0}/jars/{1}'.format(args.version, hail_jar)
+        hail_zip = 'hail-{0}-{1}.zip'.format(args.version, hail_hash)
+        py_zip = 'gs://hail-common/builds/{0}/python/{1}'.format(args.version, hail_zip)
+
+    # get Hail build (default to latest)
 
     # prepare metadata values
-    metadata = 'HASH={0},SPARK={1},HAIL_VERSION={2}'.format(hail_hash, args.spark, args.version)
+    metadata = []
+    metadata.append('JAR={}'.format(jar))
+    metadata.append('ZIP={}'.format(py_zip))
     if args.metadata:
-        metadata += ("," + args.metadata)
-
-    # if Hail jar and zip, add to metadata
-    if args.jar:
-        metadata += ',JAR={}'.format(args.jar)
-    if args.zip:
-        metadata += ',ZIP={}'.format(args.zip)
+        metadata.append(args.metadata)
 
     # if Python packages requested, add metadata variable
     if args.packages:
-        metadata = '^:^' + metadata.replace(',', ':') + ':PKGS={}'.format(args.packages)
+        metadata.append('PKGS={}'.format(args.packages.replace(',', '|')))
+
+    if args.version == 'devel':
+        metadata.append('MINICONDA_VERSION=4.4.10')
+    else:
+        metadata.append('MINICONDA_VARIANT=2')
```

Comment:
I don't quite understand this if.  MINICONDA_VERSION is controlling which version is getting installed?  We don't care for 0.1?  What's MINICONDA_VARIANT=2 do?

Reply:
MINICONDA_VARIANT=2 installs the latest miniconda for Python 2.7. MINICONDA_VARIANT=3 installs the latest miniconda for Python 3.4, because of some reasons. 🤕  
MINICONDA_VERSION is an override pointing to a specific version. In this case, a later version than MINICONDA_VARIANT=3 would give us.